Gesture events branch based on GSoC 2017 project

vadz opened the pull request:
@prashantkn94 Here is my changes mentioned in #551. They're not done yet, but should be enough for testing if destroying the gesture objects has any ill side-effects -- please let me know if it does. TIA!

My TODO list:

- [x] Collect all gesture-related data into a single struct instead of putting them in `wxWindow` directly.
- [x] Do the same thing for wxOSX.
- [x] And for wxMSW, if necessary.
- [x] Add `EnableGesturesSupport()` and only allocate this struct when needed.
- [x] Only allocate extra per-window data if it's enabled (this means using something like a map with `wxWindow*` as key and `wxWindowGestureData` as values instead of using a field in the class as we do now) in wxGTK,
- [x] And the same thing for wxOSX.
- [x] And for wxMSW too.

prashantkn94 committed:
Merge multi-touch gestures event branch

This is a squashed commit of the SOC2017_GESTURES branch from
https://github.com/prashantkn94/wxWidgets.git

Closes https://github.com/wxWidgets/wxWidgets/pull/551

vadz committed:
Make tests for gestures in wxGTK code more clear

Use wxGTK_HAS_GESTURES_SUPPORT to guard gesture-related code instead of
checking for the 3.14 GTK+ version explicitly in several places.

No real changes.

vadz committed:
Move wxGTK gesture-related data in a private opaque struct

This is a first step towards enabling gesture events only for the
windows that are interested in them as it will make it possible to avoid
wasting space on unused data in the windows that don't need it.

No real changes so far.

vadz committed:
Destroy GtkGestures objects we create

We seem to be leaking memory otherwise.

prashantkn94 commented:
I tested the changes that you have made till now and there are no ill side-effects. The behaviour of the sample is same as before.

vadz committed:
Use wxLoadedDLL for loading user32.dll dynamically

This avoids unnecessarily loading and unloading (except that it's not
really unloaded) a DLL which is always available anyhow.

vadz committed:
Slightly simplify GestureFuncs code

Document that the users of this class always must call IsOk() first, as
they already do, and remove the checks for ms_gestureSymbolsLoaded and
calls to LoadGestureSymbols() from all the other methods, which are
unnecessary in this case.

vadz committed:
Remove unnecessary assignments in WM_GESTURE handler code

The "processed" flag is false by default, no need to explicitly reset
it, this was just confusing.

vadz committed:
Use WinStruct<> to initialize GESTUREINFO

This is more concise and clear than initializing it manually.

vadz committed:
Style fixes in the MSW gestures code

No real changes, just reformat code and comments to fit in 80 column
lines.

vadz committed:
Refactor MSW gesture events initialization

Don't repeat the same code for initializing events in several different
functions but put it in a helper InitGestureEvent() function and just
call it from the event-specific handlers.

vadz committed:
Set rotation angle for the initial rotate gesture event in MSW

According to MSDN, for the initial rotation even the angle is just
passed directly, so use it.

vadz committed:
Use wxPoint instead of (x,y) pair in wxPanGestureEvent

Using higher level objects makes the code generating and using this
event shorter and more clear.

vadz committed:
Use "class" for wxGTK wxWindowGesturesData definition

No real changes, just be consistent with wxWindowGesturesData forward
declaration as "class", clang (and MSVC) warn about mismatching keywords
being used for declaration and definition.

vadz committed:
Add wxWindow::EnableTouchEvents()

Don't request touch event generation for all windows by default, this
has an inherent overhead and is not needed for 99% of the application
windows, so require calling EnableTouchEvents() explicitly to do it
instead.

Note that this requires properly initializing gesture recognizers in
wxOSX now that they're not always allocated, otherwise releasing them
when destroying the window would crash.

vadz committed:
Allow enabling individual touch gesture events

Implement support for enabling just some gesture events instead of
having to choose between getting none or all of them.

Also make wxTOUCH_NONE really disable the gestures events generation
instead of just doing nothing as before.

vadz committed:
Extract SetupCoordinates() from wxWidgetCocoaImpl

Add a free function that can be used from outside wxWidgetCocoaImpl
class too and keep the old method as a trivial wrapper around it.

No changes, this is a pure refactoring.

vadz committed:
Encapsulate gestures-related data in wxCocoaGesturesImpl

Avoid allocating a lot of almost never used data for every window and
use just a single pointer for holding all of it instead.

vadz committed:
Use global map for storing wxCocoaGesturesImpl

This decreases the speed of access to this struct (which shouldn't
matter that much) but avoids allocating an extra and almost always
unneeded pointer for each and every window, which seems like a good
trade-off.

vadz committed:
Extract "external field" approach in a reusable class

Add wxExternalField<> template that can be used for other fields and/or
in other ports too.

vadz committed:
Use wxExternalField for gestures data in wxGTK too

Don't waste memory on gesture data in all windows.

vadz committed:
Rename wxExternalField methods to not mention "Window"

This class doesn't have to be used with wxWindow or similar as object
type, so use a more neutral "Object" instead of "Window".

vadz commented:
I've finished my changes and from my point of view this should be ready to be merged now.

Unfortunately I still didn't test under any of the platforms due to lack of any multitouch hardware (actually I do have a multitouch touchpad under Linux but I have no idea what do I need to do to make it generate gesture events), so I have no idea if it still actually works or not... If you have multitouch touchpad, please do test by running the "event" sample and opening its "gestures" screen. TIA!

vadz committed:
Embed the image into the event sample

Don't load it from the image sample directory, this doesn't necessarily
work under Unix if the image sample hadn't been built.

Just embed an XPM image directly into the sample, this is good enough
for demonstration purposes.

vadz committed:
Remove hard TAB from event sample bakefile

No real changes.

nerijus commented:
I have a dualboot Android/Win10 tablet, I could test if someone gave me a compiled event sample.

e9925248 commented on include/wx/private/extfield.h:
```diff
@@ -0,0 +1,84 @@
+public:
+    typedef Object ObjectType;
+    typedef Field FieldType;
+    typedef FieldMap MapType;
```
Anything to ensure, that the map is thread safe?

vadz replied:
No, as mentioned in the comment, this is mostly meant to be used by `wxWindow`, i.e. from the GUI thread only.

e9925248 commented on interface/wx/event.h:
```diff
@@ -3644,6 +3644,251 @@ class wxScrollEvent : public wxCommandEvent
+
+    @since 3.1.1
+*/
+class wxGestureEvent : public wxEvent
```
How do you intend to integrate raw touch events into the wxEvent tree?

Each touch event has: 
* a point
* a ID to track the movement of indivual fingers
* if it is start, end or just a move [and cancel on some plattform].
* primary indicator [= considered as mouse pointer].

vadz replied:
So it looks like the current `wxGestureEvent` could be used as the base class for the new `wxMultiTouchEvent`, right? We'd just need to rename it to, say, `wxTouchEventBase` and inherit `wxMultiTouchEvent` and (now trivial) `wxGestureEvent` from it, what do you think?

e9925248 commented on src/gtk/window.cpp:
```diff
@@ -2828,6 +2893,661 @@ static gboolean source_dispatch(GSource*, GSourceFunc, void*)
+    {
+        if ( eventsMask == wxTOUCH_NONE )
+        {
+            wxWindowGestures::EraseForObject(this);
```
This will prevent disabling touch events from the event handler.

vadz replied:
Sorry, why?

e9925248 replied:
touch_callback ->  wxWindow::HandleWindowEvent -> SafelyProcessEvent -> ProcesEvent -> User-Event-Handler -> EnableTouchEvents(0);

touch_callback still uses data after returning from calling the event handler.

vadz replied:
Where does it use data after calling the event handler? I just don't see it...

e9925248 commented on src/gtk/window.cpp:
```diff
@@ -2828,6 +2893,661 @@ static gboolean source_dispatch(GSource*, GSourceFunc, void*)
+                this,
+                new wxWindowGesturesData(this, GetConnectWidget(), eventsMask)
+            );
+        }
```
It is very likely, that we will see more Options, which might not be implemented in all backend.  The current code accepts not handled bits by returning true.
If a programmer wants to enable all events specified in the mask, this behaviour is bad. Another programmer might accept, that some events can't be enabled.

What if EnableTouchEvents returns the list of enabled events instead of true/false? This way it would work for both use cases.

vadz replied:
To be honest, I think that the return value will mostly be ignored anyhow (what can you do if it fails?). But returning the mask of existing events could be useful... I really don't know if I'm going to have time to work on this myself though, but then I don't think this is critical to do before the merge.

e9925248 commented on src/gtk/window.cpp:
```diff
@@ -2828,6 +2893,661 @@ static gboolean source_dispatch(GSource*, GSourceFunc, void*)
+                wxEmitTwoFingerTapEvent(gdk_event, win);
+            }
+
+            // If the gesture was active and the touch corresponding to "press" is no longer on the screen
```
wxEmitTwoFingerTapEvent will call a used event handler, which can disable touch events. In this cases, wxWindowGesturesData is freed in the disable touch events functionl. Returning here, data is still at the old value pointing at the freed memory. The next statements will use the memory at data.

vadz replied:
Thanks, I missed this one. It could be fixed by redoing `wxWindowGestures::FromObject()` but this is definitely brittle. An obvious solution would be to use reference counting, but I wonder if it's really worth the extra complexity. And in the worst case, it's relatively trivial to use `CallAfter()` to postpone the call to `EnableTouchEvents()` slightly.

Is it really realistic to disable touch events from a touch event handler?

e9925248 replied:
Or just keep the datastructure for the whole object lifetime and delete it only in the destructor.

vadz replied:
Yes, this is probably the best/simplest solution, thanks! I'll update the PR soon.

vadz committed:
Allow safely disabling touch events from a touch event handler

Don't delete the window wxWindowGesturesData when calling
EnableTouchEvents(wxTOUCH_NONE) but just free it, to make it safe to
call EnableTouchEvents() even from a touch event handler, which later
returns to wxGTK code that could still use the gestures data object.

prashantkn94 commented:
Hi, Thank you for your work! and Please note that Touchpad gestures in GTK are available only on Wayland. and requires GDK_TOUCHPAD_GESTURE_MASK on GdkWindow to be enabled. I just tested your changes on a touchscreen laptop and I don't see any changes with the sample.

As things are getting ready to be merged, there is one issue in the sample which I want to mention here. If you do a pan gesture after zooming out then the image moves by a larger distance than it is supposed to be and same is the case with zooming in and panning but this time with a smaller distance than the normal. I have tried to fix it several times by trying different variations but with no gain so far.So, If someone with good knowledge of transformations, scaling etc is reading this then please give it a try once!.

Thanks again

vadz commented:
Ah, I've totally missed that it only worked under Wayland, thanks.

What exactly needs to be done to enable `GDK_TOUCHPAD_GESTURE_MASK`? Do we do it already and if no, why not?

prashantkn94 commented:
Adding `GDK_TOUCHPAD_GESTURE_MASK`to the list of events received by a window (`gtk_widget_add_events()`) should do the job.

I must admit that due to other tasks having higher priorities I did not worry about this one since it only requires the above thing (hopefully).

vadz committed:
Request GDK_TOUCHPAD_GESTURE_MASK events in wxGTK

This is apparently needed in order to get the gesture events.

vadz committed:
Close the gestures frame too in the event sample

Make "Exit" menu item really exit the sample again instead of just
closing the main window, but possibly leaving the gestures testing
window still open and thus not really exiting the program.

Also avoid creating more than one gesture window, use the existing one
if we had already opened it.

vadz commented:
I've added the `gtk_widget_add_events()` call, but I don't really see any difference... I still can get panning events by pressing the touchpad button and then swiping it and also long press ones, but nothing else. I've also discovered the existence of `GTK_TEST_TOUCHSCREEN`, but I have no idea how is it supposed to work, defining it doesn't seem to change anything for me neither.

vadz commented:
I have 3.22.11, but I do use X11 and not Wayland, so maybe it's just that.

a-wi commented on src/gtk/window.cpp:
```diff
@@ -3509,6 +3509,8 @@ void wxWindowGesturesData::Reinit(wxWindowGTK* win,
 
     wxASSERT_MSG( eventsMask == 0, "Unknown touch event mask bit specified" );
 
+    gtk_widget_add_events(widget, GDK_TOUCHPAD_GESTURE_MASK);
```
wxGTK_HAS_GESTURES_SUPPORT is defined for GTK+ >= 3.14 but GDK_TOUCHPAD_GESTURE_MASK is available since 3.18, hence compilation fails for GTK+ older than 3.18.

Since

vadz commented:
Thanks, I didn't notice it wasn't added together with the rest of this stuff. Should be fixed in 289fbcf5be4357cb53edd6601262d2509bb6f1f9

e9925248 commented on src/osx/cocoa/window.mm:
```diff
@@ -3416,21 +3467,21 @@ static NSComparisonResult CocoaLowerWindowCompareFunction(KindOfView first, Kind
             if ( eventsMask == wxTOUCH_NONE )
             {
-                if ( m_gesturesImpl )
+                if ( wxCocoaGesturesImpl::EraseForWindow(this) )
```
Suffers from the same freeing issue as has been fixed in GTK: Calling EnableTouchEvents from a PressAndTap Event destroys the wxCocoaGesturesImpl object, while a function call for wxCocoaGesturesImpl  is still active.

vadz replied:
Sorry, I won't have time to fix this any time soon. If you can make a PR/patch applying the same workaround, I'd be glad to apply it though. TIA!
